model : add BailingMoeV2 support

CISC opened the pull request:
Adds support for [Ling(-flash) 2.0](https://huggingface.co/collections/inclusionAI/ling-v2-68bf1dd2fc34c306c1fa6f86) and [Ring 2.0](https://huggingface.co/inclusionAI/Ring-mini-2.0), base and instruction tuned models (sans MTP, but layer included).

~Includes expert group selection implemented as a `ggml_custom` function, not sure if there's a better way, or if it makes sense to implement some sort of masking op?~

~Requires #16060 to work on CUDA, also seems to be some issue with CPU, will check later. Either way, will reimplement expert group selection with `set_rows` once possible.~

CISC committed:
add BailingMoeV2 support

CISC committed:
update llm types

CISC committed:
undo

CISC committed:
undo

CISC committed:
update llm types

CISC committed:
add model collection link

CISC committed:
update

ggerganov commented on src/llama-graph.cpp:
```diff
@@ -920,6 +977,27 @@ ggml_tensor * llm_graph_context::build_moe_ffn(
+        cb(expert_groups->src[0], "ffn_moe_group_argsort", il);
+        cb(expert_groups, "ffn_moe_group_topk", il);
+
+        // mask out the other groups
+        selection_probs = ggml_map_custom2(ctx0, selection_probs, expert_groups, mask_expert_groups, GGML_N_TASKS_MAX, (void *)(intptr_t)&hparams.n_expert_groups); // [n_expert, n_tokens]
+        cb(selection_probs, "ffn_moe_probs_masked", il);
+    }
```
We should avoid this custom op. It's good for prototyping, but it makes the implementation useless since the graph will be split in a lot of splits.

I think you can achieve this with a combination of:
```c
new_probs = ggml_scale(selection_probs, -inf);
[ some combination of repeat of expert_groups or views of the probs if needed ]
ggml_set_rows(new_probs, selection_probs, expert_groups)
```

The idea is to create a new probs vector, initialize it with -inf and then fill the selected probs in it with set_rows.

foldl replied:
FYI: Here is how I implemented this with `set_rows`. 

https://github.com/foldl/chatllm.cpp/blob/f1cf81398284e39402c06a8f6fc7c7681c6d0ff0/models/bailing.cpp#L129-L155

But, there is a missing block to bridge `top_k` and `set_rows`. This is why I created issue #16001.

ggerganov replied:
Yeah, this asymmetry in the data type of `get_rows` and `set_rows` is not great. We should resolve this.

yqy3214 commented:
Hello，I think it's necessary to split and permute the Q/K weights because of the difference in the implementation of Rotary Position Embedding (RoPE).

In the Hugging Face transformers library, the implementation is as follows:
```
def rotate_half(x):
    x1 = x[..., : x.shape[-1] // 2]
    x2 = x[..., x.shape[-1] // 2 :]
    return torch.cat((-x2, x1), dim=-1)
```
This implementation splits the features into two contiguous halves and then rotates them.

However, the implementation in llama.cpp is equivalent to this:
```
def rotate_half(x):
    """Rotates half the hidden dims of the input."""
    x1 = x[..., 0::2]
    x2 = x[..., 1::2]
    return torch.stack((-x2, x1), dim=-1).flatten(-2)
```
This version interleaves the features, selecting even and odd indices separately before rotating.

Therefore, to ensure compatibility, the model parameters for Q and K need to be split and permuted to match the llama.cpp RoPE implementation, similar to the approach in this commit: https://github.com/ggml-org/llama.cpp/pull/16028/commits/09e3df4bb5035fb3b214adfa94d539ff515f0e54

im0qianqian commented:
Hello, could we flip this PR to “Open” and put it up for review?

We're looking forward to being able to use it directly from our laptop as soon as possible. It's a lightweight model that also offers sufficiently high inference speed, and we believe it has the potential to be many people's next-generation AI assistant.

CISC commented:
> Hello, could we flip this PR to “Open” and put it up for review?

I will after I've reimplemented the expert group selection, pending the merge of #16159

CISC committed:
Merge branch 'master' into cisc/bailingmoe2

warshanks commented:
Just checking in on this, are we waiting on another PR to merge? I noticed Inclusion themselves published GGUF ahead of official support. It would be nice to get official support merged!

https://huggingface.co/inclusionAI/Ling-mini-2.0-GGUF
https://huggingface.co/inclusionAI/Ring-mini-2.0-GGUF

jacekpoplawski commented:
> Sigh, it's preferable not to release GGUFs before a PR is merged, there may be breaking changes during review...

they have their own fork for inference

CISC commented:
> Therefore, we could prioritize merging a version without group routing support. This way, at least everyone can use the model normally. Once we resolve the issue in `set_rows`, we can then submit a PR supporting group routing to unlock the model’s more outstanding performance.

I will try to resolve this today/this weekend, but as you say there really is no issue in releasing a version without it (as long as the metadata is there to support it in the future).

CISC committed:
almost working

CISC commented on src/llama-graph.cpp:
```diff
@@ -920,6 +920,29 @@ ggml_tensor * llm_graph_context::build_moe_ffn(
+        cb(expert_groups, "ffn_moe_group_topk", il);
+
+        // mask out the other groups
+        selection_probs = ggml_scale_bias(ctx0, selection_groups, 0.0f, -INFINITY);
+        group_scores = ggml_repeat_4d(ctx0, expert_groups, selection_probs->ne[1], 1, 1, 1); // [n_expert_groups, 1]
+        selection_probs = ggml_set_rows(ctx0, selection_probs, selection_groups, group_scores); // [n_tokens, n_expert_groups]
```
@ggerganov I'm having a strange issue here, the `ggml_scale_bias` does not get added to the graph, and causes a crash after `ggml_repead_4d` during `graph_compute`.

ggerganov replied:
Most likely reason is the row indices in `group_scores` are bogus for some reason.

CISC replied:
If I could achieve it the other way around it would solve the crash (though the crash feels like a bug that should be looked into), ie. if I could turn [7, 4, 0, 3] into [1, 2, 5, 6, 1, 2, 5, 6] we are home free.

ggerganov replied:
Does this fix it:

```diff
diff --git a/ggml/src/ggml.c b/ggml/src/ggml.c
index a5796214f..25580327a 100644
--- a/ggml/src/ggml.c
+++ b/ggml/src/ggml.c
@@ -3687,6 +3687,7 @@ struct ggml_tensor * ggml_set_rows(
     result->op     = GGML_OP_SET_ROWS;
     result->src[0] = b;
     result->src[1] = c;
+    result->src[2] = a;
 
     return result;
 }
```

CISC replied:
> Does this fix it:

Yep, that did it! :tada:

CISC replied:
Yay, another bug, `ggml_scale_bias(a, 0.0f, -INFINITE)` results in -nan on CPU.

CISC committed:
Merge branch 'master' into cisc/bailingmoe2

CISC commented on src/llama-hparams.h:
```diff
@@ -72,6 +72,8 @@ struct llama_hparams {
     uint32_t n_expert_shared    = 0;
     uint32_t n_norm_groups      = 0;
+    uint32_t n_expert_groups    = 0;
+    uint32_t n_group_exp        = 0;
     uint32_t n_group_experts    = 0;
```
Ok, these probably need to be renamed...

CISC replied:
I guess `n_group_exp` should be renamed to `n_group_used` as it's basically `n_expert_used` for groups.

CISC commented on src/llama-graph.cpp:
```diff
@@ -928,6 +928,29 @@ ggml_tensor * llm_graph_context::build_moe_ffn(
+        const int64_t n_exp_per_group = n_expert / hparams.n_expert_groups;
+
+        // organize experts into n_expert_groups
+        ggml_tensor * selection_groups = ggml_view_2d(ctx0, ggml_cont(ctx0, ggml_transpose(ctx0, selection_probs)), n_tokens * n_exp_per_group, hparams.n_expert_groups, n_tokens * n_exp_per_group * sizeof(float), 0); // [n_tokens, n_expert_groups]
```
The shapes are right, the thinking in the comment was just that it's the combined `n_tokens` for each `n_expert_group`.

The logic is merely that we need to process the experts grouped together instead of individually to figure out which group scores highest.

CISC commented on src/llama-graph.cpp:
```diff
@@ -928,6 +928,29 @@ ggml_tensor * llm_graph_context::build_moe_ffn(
+        ggml_tensor * group_scores = ggml_top_k(ctx0, selection_groups, 2); // [2, n_expert_groups]
+
+        // get top n_group_exp expert groups
+        group_scores = ggml_transpose(ctx0, ggml_sum_rows(ctx0, ggml_cast(ctx0, group_scores, GGML_TYPE_F32))); // [n_expert_groups, 1]
```
Ooops, yep, forgot a `get_rows` there.

CISC commented on src/llama-graph.cpp:
```diff
@@ -928,6 +928,29 @@ ggml_tensor * llm_graph_context::build_moe_ffn(
+
+        // organize experts into n_expert_groups
+        ggml_tensor * selection_groups = ggml_view_2d(ctx0, ggml_cont(ctx0, ggml_transpose(ctx0, selection_probs)), n_tokens * n_exp_per_group, hparams.n_expert_groups, n_tokens * n_exp_per_group * sizeof(float), 0); // [n_tokens, n_expert_groups]
+        ggml_tensor * group_scores = ggml_top_k(ctx0, selection_groups, 2); // [2, n_expert_groups]
+
+        // get top n_group_exp expert groups
+        group_scores = ggml_transpose(ctx0, ggml_sum_rows(ctx0, ggml_cast(ctx0, group_scores, GGML_TYPE_F32))); // [n_expert_groups, 1]
```
```suggestion
        ggml_tensor * group_scores = ggml_top_k(ctx0, selection_groups, 2); // [2, n_expert_groups]
        group_scores = ggml_get_rows(ctx0, ggml_reshape_3d(ctx0, selection_groups, 1, selection_groups->ne[0], selection_groups->ne[1]), group_scores); // [1, 2, n_expert_groups]

        // get top n_group_exp expert groups
        group_scores = ggml_transpose(ctx0, ggml_sum_rows(ctx0, ggml_reshape_2d(ctx0, group_scores, group_scores->ne[1], group_scores->ne[2]))); // [n_expert_groups, 1]
```

CISC committed:
correct group selection and rename n_group_exp

CISC commented:
Sigh, this is cursed, with certain prompts I get the following error:
```
ggml_cuda_compute_forward: ARGSORT failed
CUDA error: invalid argument
```
and I can't figure out what it errors on, I've thrown in tons of asserts to no avail.

@JohannesGaessler any tips to figure this out?

CISC commented:
> In `argsort.cu`:
> 
> ```c
> const dim3 block_dims(ncols_pad, 1, 1);
> const dim3 block_nums(1, nrows, 1);
> ```
> 
> If nrows is ever >= 65536 the kernel launch will fail because the CUDA grid in dimensions y and z only uses 16 bits (the x dimension uses 32 bits). Try simply putting `nrows` in the x dimension and swap `blockIdx.y` with `blockIdx.x`.

Actually, it looks like it's `block_dims` that's the issue, if I force it to f.ex. `256` all is well.

CISC commented:
@JohannesGaessler As far as I understand CUDA argsort it can't actually handle more than N-threads-per-block columns (which is not checked, hence the `invalid arguments`). It also has a limit on columns depending on shared mem, but the thread limit hits first.

slaren commented:
The argsort kernel was made for MoE, where the limits are never hit. For bigger tensors, an option could be to use cub.

slaren commented:
For now, modifying the `supports_op` to check limits should be enough to avoid the crash, an optimization can come later. Btw the CPU implementation is also very bad for big tensors (it's a bubble sort).

CISC committed:
avoid large top_k and use argmax instead for now

if we had something like argmax2 that would be equivalent, but this works fine until then

CISC committed:
poke

CISC committed:
Merge branch 'master' into cisc/bailingmoe2

CISC committed:
Merge branch 'master' into cisc/bailingmoe2

CISC committed:
Merge branch 'master' into cisc/bailingmoe2

CISC committed:
skip group selection when there are no tokens

CISC commented:
@ikawrakow Fixed, thanks again for the report.

CISC committed:
fix 1T conversion

CISC committed:
hopefully fixed expert group selection

third time's the charm?

ubergarm commented:
@CISC 

Thanks for your help, I had time to roll a mainline quant compatible Ling-1T test quant `smol-IQ2_XXS` coming in at `249.92 GiB (2.15 BPW)` so should fit on a 256GB RAM+24GB VRAM rig.

Early tests with this PR `llama.cpp/cisc/bailingmoe2@bc6c48af1` are successful compiling CPU-only backend and running like so:

<details>

<summary>👈 Details</summary>

```
model=/mnt/data/models/ubergarm/Ling-1T-GGUF/Ling-1T-smol-IQ2_XXS.gguf
numactl -N "$SOCKET" -m "$SOCKET" \
./build/bin/llama-server \
    --model "$model"\
    --alias ubergarm/Ling-1T-smol-IQ2_XXS \
    --ctx-size 65536 \
    -fa 1 \
    -ctk q8_0 -ctv q8_0 \
    -ub 4096 -b 4096 \
    --parallel 1 \
    --threads 128 \
    --threads-batch 192 \
    --numa numactl \
    --host 127.0.0.1 \
    --port 8080 \
    --no-mmap

llama_model_loader: - type  f32:  473 tensors
llama_model_loader: - type q8_0:    8 tensors
llama_model_loader: - type q4_K:  161 tensors
llama_model_loader: - type q5_K:   80 tensors
llama_model_loader: - type q6_K:  153 tensors
llama_model_loader: - type iq2_xxs:  228 tensors
```

</details>

With some luck I'll be able to clear up my huggingface `403` errors and upload this test quant for people to try over at: https://huggingface.co/ubergarm/Ling-1T-GGUF

I can provide the exact recipe if needed just holler.

Thanks and great work and appreciate all the collaboration!

CISC committed:
make expert group selection generally available

The new LLaDA2Moe model uses this method too, make it generally available regardless of architecture.

CISC committed:
allow n_expert_groups to be 1 (Kimi K2) [no ci]

ggerganov commented on src/llama-graph.cpp:
```diff
@@ -928,6 +928,31 @@ ggml_tensor * llm_graph_context::build_moe_ffn(
     }
 
+    // select top n_group_used expert groups
+    if (hparams.n_expert_groups > 1 && n_tokens > 0) {
```
Add link to the reference implementation

ggerganov reviewed (approved):
The expert grouping logic seems correct to me. But ideally you should compare the logits with the reference implementation just in case.

CISC committed:
address review suggestions
